postgres_insert_block.py

```diff
@@ -1,61 +1,108 @@
+                                          "not be executed")
+                    self._rollback_transactions()
+            else:
+                # commit after every insert
+                try:
+                    for signal in signals:
+                        self.execute_insert(signal.to_dict())
+                        self._commit_transactions()
+                except:
+                    self.logger.exception("Insert query could not be executed")
                     self._rollback_transactions()
 
     def execute_insert(self, data):
         """execute an insert query for the given data"""
         self.logger.debug('executing INSERT on data: {}'.format(data))
 
         # first check to see if the keys are valid
-        for key in data.keys():
-            self._validate_string(key)
+        if self.bulk_insert():
+            # grab the keys of the first signal
+            for key in data[0].keys():
+                self._validate_string(key)
+        else:
+            for key in data.keys():
+                self._validate_string(key)
 
-        self._cur.execute(self._build_insert_query_string(data),
-                          tuple(data.values()))
+        # execute the insert query. the query string is built with the data
+        # already in it, so jsut execute that query here.
+        self._cur.execute(self._build_insert_query_string(data))
 
     def _build_insert_query_string(self, data):
-        """build an INSERT SQL query based on the incoming (dictionary)
-        data. psycopg2 automatically escapes/sanitizes values passed into
-        execute and mogrify. First mogrifying the table name and data keys
-        assures escaping of those variables, and the data values will be
-        escaped when execute is called.
+        """build an INSERT SQL query based on the incoming data. the keys are
+        validated before building this query, so just insert those into the
+        query. Then build the value tuple(s), which are mogrified/escaped
+        correctly while building the string.
         """
 
-        query_base = 'INSERT INTO {} ({}) VALUES ({})'
-        query_final = query_base.format(self.table_name(),
-                                        ', '.join(list(data.keys())),
-                                        ', '.join(['%s'] * len(data))
-                                        )
+        if isinstance(data, list):
+            # this is a bulk insert. Take the keys from the first signal and
+            # assume that those keys will be the same for all the insert
+            # signals, since we're only inserting into one table. Then make
+            # n tuples of values where n is the number of signals.
+            keys_tuple = ', '.join(list(data[0].keys()))
+            value_tuple = ','.join(
+                self._cur.mogrify(
+                    "({})".format(','.join(["%s"] * len(data_dict))),
+                    tuple(data_dict.values())).decode()
```

Comment:
I still think we need to have some kind of `sort` on this. What if two signals in the list have the same attributes but `values()` returns them in a different order?